File: Golang/692. Top K Frequent Words.go
// TC: O(nlgn) SC: O(n)
import "sort"

func topKFrequent(words []string, k int) []string {
	var set []string
	freq := make(map[string]int)

	for _, w := range words {
		freq[w]++
		if freq[w] == 1 {
			set = append(set, w)
		}
	}
	sort.Slice(set, func(i, j int) bool {
		if freq[set[i]] == freq[set[j]] {
			return set[i] < set[j]
		}
		return freq[set[i]] > freq[set[j]]
	})
	return set[:k]
}

// Input: words = ["i","love","leetcode","i","love","coding"], k = 2
// Output: ["i","love"]

File: Python/692. Top K Frequent Words.py
# O(Nlogk), O(n)
import heapq
import collections
from typing import List

class Solution:
    def topKFrequent(self, words: List[str], k: int) -> List[str]:
        count = collections.Counter(words) # {'i': 2, 'love': 2, 'leetcode': 1, 'coding': 1}
        heap = [(-freq, word) for word, freq in count.items()]
        heapq.heapify(heap)
        # [(-2, 'i'), (-2, 'love'), (-1, 'leetcode'), (-1, 'coding')]
        # print(f"count = {count}, heap={heap}")
        return [heapq.heappop(heap)[1] for _ in range(k)]
        